src/drivers/roboclaw/RoboClaw.hpp

```diff
@@ -0,0 +1,298 @@
+	static bool taskShouldExit;
+
+	/** control channels */
+	enum e_channel {
+		CH_VOLTAGE_LEFT = 0,
+		CH_VOLTAGE_RIGHT
+	};
+
+	/**  motors */
+	enum e_motor {
+		MOTOR_1 = 0,
+		MOTOR_2
+	};
+
+	/** error handeling*/
+	enum class RoboClawError {
+		Success,
+		WriteError,
+		ReadError,
+		ChecksumError,
+		ChecksumMismatch,
+		UnexpectedError,
+		ReadTimeout
+	};
+
+	/**
+	 * constructor
+	 * @param deviceName the name of the
+	 * 	serial port e.g. "/dev/ttyS2"
+	 * @param address the adddress  of the motor
+	 * 	(selectable on roboclaw)
+	 * @param baudRateParam Name of the parameter that holds the baud rate of this serial port
+	 */
+	RoboClaw(const char *deviceName, const char *baudRateParam);
+
+	/**
+	 * deconstructor
+	 */
+	virtual ~RoboClaw();
+
+	void initialize();
+
+	/**
+	 * @return position of a motor, rev
+	 */
+	float getMotorPosition(e_motor motor);
+
+	/**
+	 * @return speed of a motor, rev/sec
+	 */
+	float getMotorSpeed(e_motor motor);
+
+	/**
+	 * set the speed of a motor, rev/sec
+	 */
+	int setMotorSpeed(e_motor motor, float value);
+
+	/**
+	 * set the duty cycle of a motor
+	 */
+	int setMotorDutyCycle(e_motor motor, float value);
+
+	/**
+	 * Drive both motors. +1 = full forward, -1 = full backward
+	 */
+	int drive(float value);
+
+	/**
+	 * Turn. +1 = full right, -1 = full left
+	 */
+	int turn(float value);
+
+	/**
+	 * reset the encoders
+	 * @return status
+	 */
+	int resetEncoders();
+
+	/**
+	 * read data from serial
+	 */
+	int readEncoder();
+
+	/**
+	 * print status
+	 */
+	void printStatus();
+
+
+private:
+
+	char _storedDeviceName[256]; // Adjust size as necessary
+    	char _storedBaudRateParam[256]; // Adjust size as necessary
+
+	int _timeout_counter = 0;
+
+	// commands
+	// We just list the commands we want from the manual here.
+	enum e_command {
+
+		// simple
+		CMD_DRIVE_FWD_1 = 0,
+		CMD_DRIVE_REV_1 = 1,
+		CMD_DRIVE_FWD_2 = 4,
+		CMD_DRIVE_REV_2 = 5,
+
+		CMD_DRIVE_FWD_MIX = 8,
+		CMD_DRIVE_REV_MIX = 9,
+		CMD_TURN_RIGHT = 10,
+		CMD_TURN_LEFT = 11,
+
+		// encoder commands
+		CMD_READ_ENCODER_1 = 16,
+		CMD_READ_ENCODER_2 = 17,
+		CMD_READ_SPEED_1 = 18,
+		CMD_READ_SPEED_2 = 19,
+		CMD_RESET_ENCODERS = 20,
+		CMD_READ_BOTH_ENCODERS = 78,
+		CMD_READ_BOTH_SPEEDS = 79,
+
+		// advanced motor control
+		CMD_READ_SPEED_HIRES_1 = 30,
+		CMD_READ_SPEED_HIRES_2 = 31,
+		CMD_SIGNED_DUTYCYCLE_1 = 32,
+		CMD_SIGNED_DUTYCYCLE_2 = 33,
+
+		CMD_READ_STATUS = 90
+	};
+
+
+	struct {
+		speed_t serial_baud_rate;
+		int32_t counts_per_rev;
+		int32_t encoder_read_period_ms;
+		int32_t actuator_write_period_ms;
+		int32_t address;
+	} _parameters{};
+
+	struct {
+		param_t serial_baud_rate;
+		param_t counts_per_rev;
+		param_t encoder_read_period_ms;
+		param_t actuator_write_period_ms;
+		param_t address;
+	} _param_handles{};
+
+	int _uart;
+	fd_set _uart_set;
+	struct timeval _uart_timeout;
+
+	/** actuator controls subscription */
+	int _actuatorsSub{-1};
+	actuator_controls_s _actuatorControls;
```

Comment:
```suggestion
```

All of this can go and be replaced with the OutputModuleInterface. https://github.com/PX4/PX4-Autopilot/blob/d83b9f3c38a372591b9389507feaa53111f209e0/src/lib/mixer_module/mixer_module.hpp#L69-L93

I can explain later if it isn't clear, but looking at an example like drivers/dshot or drivers/pwm_out might help.

https://github.com/PX4/PX4-Autopilot/blob/d83b9f3c38a372591b9389507feaa53111f209e0/src/drivers/dshot/DShot.cpp#L339-L340

https://github.com/PX4/PX4-Autopilot/blob/d83b9f3c38a372591b9389507feaa53111f209e0/src/drivers/dshot/DShot.h#L146

Reply:
This was done in 50ba028cc513237e50e29d7cbca0852e21413ca4